src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/Compression/WebSocketDeflater.cs

```diff
@@ -0,0 +1,229 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using static System.IO.Compression.ZLibNative;
+
+namespace System.Net.WebSockets.Compression
+{
+    /// <summary>
+    /// Provides a wrapper around the ZLib compression API.
+    /// </summary>
+    internal sealed class WebSocketDeflater : IDisposable
+    {
+        private ZLibStreamHandle? _stream;
+        private readonly int _windowBits;
+        private readonly bool _persisted;
+
+        private byte[]? _buffer;
+
+        internal WebSocketDeflater(int windowBits, bool persisted)
+        {
+            Debug.Assert(windowBits >= 9 && windowBits <= 15);
+
+            // We use negative window bits in order to produce raw deflate data
+            _windowBits = -windowBits;
+            _persisted = persisted;
+        }
+
+        public void Dispose() => _stream?.Dispose();
+
+        public void ReleaseBuffer()
```

Comment:
Good question. @davidfowl basically asked the same here https://github.com/dotnet/runtime/pull/49304#discussion_r593810980. We need to keep a reference to this buffer while the ManagedWebSocket copies the deflated bytes into its _sendBuffer. Then it calls ReleaseBuffer. This can be avoided and be made more explicit if I returned IMemoryOwner<byte> from the Deflate() method and not ReadOnlySpan<byte> as it's now, but this will cause memory allocations for each send for the instance implementing IMemory<Owner>. 

The compromise is to return a ReadOnlySpan<byte> so the implementation in ManagedWebSocket for the send is not affected in either case (compressed or not), only that it needs to call ReleaseBuffer when _deflater is not null - see https://github.com/zlatanov/runtime/blob/32013aad69cb967b0359640b6a9be6b54205d71f/src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs#L539-L586.

The inflater doesn't need to expose it's buffer because ManagedWebSocket doesn't care about it, it delegates the `Read` operation to the inflater.

Reply:
I've looked at what chromium does, and indeed it uses what you suggest. See https://github.com/chromium/chromium/blob/99314be8152e688bafbbf9a615536bdbb289ea87/net/websockets/websocket_deflater.cc#L132-L136.

I guess in the case when the websocket is not a server, we can opt into that and in the case of a server websocket, keep the current implementation.